src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,414 @@
+        /// Element-wise conversions occur late, together with the element-wise comparisons. They may not be evaluated.
+        /// </summary>
+        /// There are 2 cases:
+        /// - tuple literal cannot be null and it produces as many temps as elements
+        /// - tuple types or nullable tuple types produce a single temp (and no further temps are initialized)
+        public override BoundNode VisitTupleBinaryOperator(BoundTupleBinaryOperator node)
+        {
+            var boolType = node.Type; // we can re-use the bool type
+            var initialEffectsAndTemps = TupleOperatorSideEffectsAndTemps.GetInstance();
+
+            var returnValue = RewriteTupleOperator(node.Operators, node.Left, node.Right, boolType, initialEffectsAndTemps, node.OperatorKind, inLeftLiteral: true, inRightLiteral: true);
+
+            var (effects, temps) = initialEffectsAndTemps.ToImmutableAndFree();
+            var result = _factory.Sequence(temps, effects, returnValue);
+            return result;
+        }
+
+        private BoundExpression RewriteTupleOperator(TupleBinaryOperatorInfo @operator,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType,
+            TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftLiteral, bool inRightLiteral)
+        {
+            if (@operator.IsSingle())
+            {
+                return RewriteTupleSingleOperator((TupleBinaryOperatorInfo.Single)@operator, left, right, boolType, operatorKind);
+            }
+            else
+            {
+                return RewriteTupleNestedOperators((TupleBinaryOperatorInfo.Multiple)@operator, left, right, boolType, initialEffectsAndTemps, operatorKind, inLeftLiteral, inRightLiteral);
+            }
+        }
+
+        private BoundExpression RewriteTupleNestedOperators(TupleBinaryOperatorInfo.Multiple operators, BoundExpression left, BoundExpression right,
+            TypeSymbol boolType, TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftLiteral, bool inRightLiteral)
+        {
+            // If either left or right is nullable, produce:
+            //
+            //      // outer sequence
+            //      var leftHasValue = left.HasValue; (or true if !leftNullable)
+            //      var rightHasValue = right.HasValue; (or true if !rightNullable)
+            //      leftHasValue == rightHasValue
+            //          ? leftHasValue ? ... inner sequence ... : true/false
+            //          : false/true
+            //
+            // where inner sequence is:
+            //      var leftValue = left.GetValueOrDefault(); (or left if !leftNullable)
+            //      var rightValue = right.GetValueOrDefault(); (or right if !rightNullable)
+            //      ... logical expression using leftValue and rightValue ...
+            //
+            // and true/false and false/true depend on operatorKind (== vs. !=)
```

Comment:
>// and true/false and false/true depend on operatorKind (== vs. !=) [](start = 12, length = 67)

Consider if we can apply optimizations similar to TrivialLiftedComparisonOperatorOptimizations

Reply:
Added PROTOTYPE

---
In reply to: [171075510](https://github.com/dotnet/roslyn/pull/23356#discussion_r171075510) [](ancestors = 171075510)